robotoff/prediction/ocr/origin.py

```diff
@@ -0,0 +1,264 @@
+    Utility function for checking if one of the general words like 'elements' or 'ingredients' is 
+    present in the list of ingredients returned by extract_ingredients_from_match_string
+    """
+    for reg in general_words_regex:
+        if re.search(reg, ingredients) is not None:
+            return True
+    return False
+
+def extract_origin_from_match (origin_match, lang: str) -> str:
+    if origin_match.group("in_or_outside") is not None and origin_match.group("in_or_outside") in OUTSIDE_WORDS_BY_LANG[lang]:
+        return LARGE_ORIGIN # the sentence tells that the product is made outside a zone, 
+        # so the only thing we can extract is that it has a large origin
+    elif origin_match.group("negation") is not None:
+        return LARGE_ORIGIN # the sentence tells that the product is not made in the specified zone, 
+        # so the only thing we can extract is that it has a large origin
+    elif origin_match.group("country") is not None:
+        def standardize (s):
+            return " ".join((word[0].upper() + word[1:].lower() for word in s.split(" ")))
+        standardized_name = standardize(origin_match.group("country"))
+        try:
+            return next( #find the translation in english of the specified country name 
+                country_id
+                for country_id, country_content in COUNTRIES_IN_ALL_LANGS.items()
+                if country_content["name"][lang] == standardized_name
+            )
+        except StopIteration: #impossible, if the regex matched with the group country, the country with this name exists
+            return standardized_name
+    elif origin_match.group("country_adj") is not None and origin_match.group("country") in COUNTRY_ADJECTIVE_TO_COUNTRY_NAME_BY_LANG[lang]:
+        return COUNTRY_ADJECTIVE_TO_COUNTRY_NAME_BY_LANG[lang][origin_match.group("country_adj").lower()]
+    elif origin_match.group("large_origin") is not None:
+        return LARGE_ORIGIN
+    else:
+        return UNKNOW_ORIGIN
+
+def extract_ingredients_from_match_string (ingredients_text: str, lang) -> List[str]:
+    """
+    Utility function that splits the regex's matched group "ingredient" into a list of ingredients.
+    the regex matches with sentence with several ingredients, like 'quinoa and rice are made in France'
+    It match the whole "quinoa and rice " string, and this function splits each ingredients.
+    """
+    splitted = ingredients_text.split(" ")
+    def is_not_empty (s):
+        return s != ""
+    splitted = filter(is_not_empty, splitted)
+    def standardize_ingredient (s):
+        return s[0].upper() + s[1:].lower()
+    standardized = map(standardize_ingredient, splitted) 
+    def to_id (s):
+        try:
+            return next(ingredient_id for ingredient_id, ingredient in INGREDIENTS.items() 
+                if lang in ingredient["name"]
+                and ingredient["name"][lang] == s
+            )
+        except StopIteration:
+            return s
```

Comment:
As we know we will do this kind of lookup, I would build a dict at initialization time, so that we can just check the key directly, it will really improve performance.
So that I can do 
```python
return INGREDIENTS_ID[lang].get(s, s)
```

building the INGREDIENTS_ID is something like:
```python
INGREDIENTS_ID = collections.defaultdict(dict)
for ingredient_id, ingredient in INGREDIENTS.items():
    for lang, name in ingredient["name"].items():
        INGREDIENTS_ID[lang][name] = ingredient_id
````

Reply:
Good idea, I'll do this while making all this a class